### DISCUSSION TITLE
Using Multiple Tags with Arduino

### DISCUSSION THREAD

**Discussion by HardX18** [bitcraze/discussions#1281] [Ideas]
Can we use multiple tags using the libdw1000 library. In which the UWB module is used with the Arduino, may be using TDoA protocol ??

As it may be helpful in implementing the protocol in other application.

**Comment by ataffanel** [comment]
As mentioned in the driver readme, the driver is designed to be able to control multiple UWB module from a single MCU: https://github.com/bitcraze/libdw1000?tab=readme-ov-file#initialize. All you should have to do is to make a separate ```dwDevice_t``` and ```dwOps_t``` for each DWM module. This will allow you to control each module independently.

**Comment by HardX18** [comment]
This isn't exactly what I'm inquiring about. I'm proposing the possibility of utilizing various UWB modules with different controllers as tags, alongside having N number of anchors in the TDoA protocol, similar to what's employed in the Crazyflie swarm.

If this configuration is feasible, it could lead to the creation of various systems applicable to our research project needs.

While I was experimenting with the TWR protocol, I found that I could only have one tag functioning with four anchors.

Additionally, could you please provide some sample code that can be utilized with Arduino for testing the TDoA protocol ?

**Reply by ataffanel** [reply]
I do not understand the question, could you maybe add more context and precision? What UWB modules and what controllers are you referring to?

Libdw1000 gives low level access to the DW1000 and allows to implement any algorithm on top of it. We have never implemented TDoA on Arduino so we do not have any example of it, though the Crazyflie source code is written in C so it should be possible to port TDoA to the Arduino platform.

**Comment by HardX18** [comment]
Hi,

Yes, I was able to port the library (libdw1000) to Arduino.

Regarding the context - I'm attempting to establish the TDoA protocol using ESP32 and UWB. Since the existing TWR protocol only allows for one tag to be used at a time, but I need to accommodate multiple tags, that why I'm exploring TDoA.

Could you please provide any example code where you have just implemented TDoA with any system, perhaps in the testing phase? This would give me an idea of how to proceed with the implementation.

**Reply by ataffanel** [reply]
You can find the current TDoA implementation in the Crazyflie there: https://github.com/bitcraze/crazyflie-firmware/tree/master/src/utils/src/tdoa. If you want earlier version you can look at the Crazyflie history in Git and at the Crazyflie documentation (we have 2 flavour of TDoA, TDoA2 that supports 8 anchor and TDoA3 that is asynchronous and dynamically adapt to any number of anchor).

This should be all I can help with since this is a bit out of scope of this forum; we can help for things that relates to the Crazyflie ecosystem, but not so much for other projects.

**Comment by HardX18** [comment]
Hi,

Thank you for your response. I've already visited the link you provided. However, I'm having difficulty understanding when and where those functions are being called, as the TDoA Engine file you mentioned only contains function definitions.

Could you please assist me with a scenario? If I have eight crazyflie anchors and tag is of ESP+UWB, which function should be called to determine the position of the tag? I'm assuming that I'm using the Crazyflie Nodes with the LPS firmware. Thank you!"

**Reply by knmcguire** [reply]
I'll second on @ataffanel on this one. We have quite a lot of documentation available on how TDOA works and how it has been implemented in code as it is opensource. We can't however, help out for hardware that we haven't designed ourselves as we would require the exact same hardware as well. So we will go as far as point you to the documentation but I'm afraid that the bulk of the work needs to be done by yourself:

- https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/loco-positioning-system/
- https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/

I'd also like to refer you to [our forum rules](https://github.com/orgs/bitcraze/discussions/55), and[ another discussion](https://github.com/orgs/bitcraze/discussions/1037) where we have explained to another user why we can only have limited support with self-made hardware.

**Comment by HardX18** [comment]
Thank you for the reply,

I have gone through the links you have provided and I have also ported the code for the ESP controller. I am asking your help if you can provide any example code which you might have used with the crazyflie hardware which is just to get the location from the crazyflie anchor (code when you were testing the implementation of UWB deck at the starting) as that would be a good starting point for me and would help me to make an opensource project.

**Comment by HardX18** [comment]
hi,

can you please provide me the testing code you have used ??

**Reply by ataffanel** [reply]
We cannot do that since we do not have such testing code. We worked directly on the Crazyflie and all we have from this work is in the Crazyflie source tree, you can look at the git history if you want to see some earlier code.
